x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/async/AsyncTaskIndexService.java

```diff
@@ -356,78 +329,116 @@ private void getEncodedResponse(AsyncExecutionId asyncExecutionId,
     public void getResponse(AsyncExecutionId asyncExecutionId,
                             boolean restoreResponseHeaders,
                             ActionListener<R> listener) {
-        getEncodedResponse(asyncExecutionId, restoreResponseHeaders, ActionListener.wrap(
-            (t) -> listener.onResponse(decodeResponse(t.v1()).withExpirationTime(t.v2())),
-            listener::onFailure
-        ));
+        getResponseFromIndex(asyncExecutionId, restoreResponseHeaders, true, listener);
+    }
+
+    private void getResponseFromIndex(AsyncExecutionId asyncExecutionId,
+                                      boolean restoreResponseHeaders,
+                                      boolean requireSameUser,
+                                      ActionListener<R> outerListener) {
+        final GetRequest getRequest = new GetRequest(index)
+            .preference(asyncExecutionId.getEncoded())
+            .id(asyncExecutionId.getDocId())
+            .realtime(true);
+        clientWithOrigin.get(getRequest, outerListener.delegateFailure((listener, getResponse) -> {
+            if (getResponse.isExists() == false) {
+                listener.onFailure(new ResourceNotFoundException(asyncExecutionId.getEncoded()));
+                return;
+            }
+            long reservedBytes = 0;
+            // Parse the source manually so we can access the encoded buffer directly without making it a string
+            try (XContentParser parser = XContentHelper.createParser(NamedXContentRegistry.EMPTY,
+                DeprecationHandler.THROW_UNSUPPORTED_OPERATION, getResponse.getSourceInternal(), XContentType.JSON)) {
+                ensureExpectedToken(parser.nextToken(), XContentParser.Token.START_OBJECT, parser);
+                R resp = null;
+                Long expirationTime = null;
+                while (parser.nextToken() != XContentParser.Token.END_OBJECT) {
+                    ensureExpectedToken(XContentParser.Token.FIELD_NAME, parser.currentToken(), parser);
+                    final String fieldName = parser.currentName();
+                    parser.nextToken();
+                    switch (fieldName) {
+                        case RESULT_FIELD:
+                            ensureExpectedToken(XContentParser.Token.VALUE_STRING, parser.currentToken(), parser);
+                            final CharBuffer encodedBuffer = parser.charBuffer();
```

Comment:
Do you think we can use `parser.binaryValue()` instead that will provide already decoded from Base64 array?

Reply:
I looked at the code of `getBinaryValue` and `readBinaryValue` of JsonParser. Both use two intermediate buffers while decoding a Base64 string. The current approach doesn't require extra memory. I know we shouldn't utilize these optimizations, but they help when encoding/decoding huge responses.

Reply:
Thanks for the explanation, Nhat.

I've checked `parser.charBuffer` implementation, it uses `String::toCharArray()` which will allocate 
a newly character array whose length is the length of this string.   So it means that we first allocate 2X Response size  memory, and only use circuit breaker for another 1X memory?

Reply:
I had some other ideas, what do you think of it:
1) You said we can record the size of search response, what if we record the size of just encoded Based 64 string (which we should know)
2) Another very simple way is to use setting `search.max_async_search_response_size` that #74455 introduces.  We can just check for max possible size available in memory.

Reply:
> I've checked parser.charBuffer implementation, it uses String::toCharArray() which will allocate
a newly character array whose length is the length of this string.

Great catch, thanks Mayya! I have updated this PR to account an extra buffer to the XContentParser. Unlike the reserved memory for the response, we can release this memory immediately after parsing.

> You said we can record the size of search response, what if we record the size of just encoded Based 64 string (which we should know)

We already have it when parsing the xContent (i.e., encodedBuffer.length()).

> Another very simple way is to use setting search.max_async_search_response_size that Set max allowed size for stored async response #74455 introduces. We can just check for max possible size available in memory.

We can overly reserve the memory especially when users use a large value for this setting.